File: src/server/services/chatService.ts
import dotenv from "dotenv"
import { db } from "../../db/index.js"
import {
  recipesTable,
  recipeIngredientsTable,
  ingredientsTable,
} from "../../db/schema.js"
import { eq } from "drizzle-orm"
import { sql } from "drizzle-orm"
import OpenAI from "openai"

dotenv.config()

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
})

interface Message {
  role: "system" | "user" | "assistant"
  content: string
}

interface RecipeWithIngredients {
  id: number
  name: string
  description: string | null
  ingredients: Array<{ name: string }>
}

interface ChatState {
  messages: Message[]
  previouslySuggestedRecipes: number[] // Track recipe IDs that were already suggested
}

const chatStates = new Map<string, ChatState>()

// Map common food terms to related ingredients and cuisines
const foodRelations = new Map<string, string[]>([
  [
    "pizza",
    [
      "italian",
      "tomato",
      "cheese",
      "basil",
      "dough",
      "pepperoni",
      "mozzarella",
    ],
  ],
  ["pasta", ["italian", "tomato", "garlic", "parmesan", "basil", "olive oil"]],
  ["burger", ["beef", "bun", "cheese", "lettuce", "tomato", "onion"]],
  [
    "salad",
    ["lettuce", "tomato", "cucumber", "olive oil", "vinegar", "healthy"],
  ],
  ["steak", ["beef", "garlic", "butter", "rosemary", "thyme"]],
  ["chicken", ["poultry", "garlic", "herbs", "lemon", "olive oil"]],
  ["fish", ["seafood", "lemon", "butter", "herbs", "garlic"]],
  ["soup", ["broth", "vegetables", "herbs", "warm", "comfort"]],
  ["stir fry", ["asian", "soy sauce", "vegetables", "ginger", "garlic"]],
  ["curry", ["indian", "spices", "coconut", "rice", "ginger", "garlic"]],
  ["tacos", ["mexican", "tortilla", "salsa", "cilantro", "lime"]],
  ["sushi", ["japanese", "rice", "seafood", "nori", "wasabi"]],
])

// List of cuisine types to help identify food preferences
const cuisineTypes = new Set([
  "italian",
  "mexican",
  "chinese",
  "japanese",
  "indian",
  "thai",
  "mediterranean",
  "french",
  "american",
  "korean",
  "vietnamese",
  "greek",
  "spanish",
  "middle eastern",
])

async function extractKeywordsWithAI(userInput: string): Promise<string[]> {
  try {
    console.log("Extracting keywords from user input with OpenAI:", userInput)

    const response = await openai.chat.completions.create({
      model: "gpt-3.5-turbo",
      messages: [
        {
          role: "system",
          content: `You are a culinary expert. Extract up to 10 food-related keywords from the user's input. 
          Focus on ingredients, cuisines, dishes, and cooking methods. 
          Respond ONLY with a comma-separated list of lowercase keywords, no explanations.
          Example input: "I love spicy Asian food and seafood"
          Example response: "spicy, asian, seafood, stir-fry, noodles, curry"`,
        },
        {
          role: "user",
          content: userInput,
        },
      ],
      temperature: 0.3,
      max_tokens: 50,
    })

    const keywords =
      response.choices[0]?.message?.content
        ?.split(",")
        .map((keyword) => keyword.trim().toLowerCase())
        .filter((keyword) => keyword.length > 0) || []

    console.log("OpenAI extracted keywords:", keywords)
    return keywords
  } catch (error) {
    console.error("Error extracting keywords with OpenAI:", error)
    // Fall back to basic keyword extraction if OpenAI fails
    return extractFoodKeywords(userInput)
  }
}

function extractFoodKeywords(input: string): string[] {
  const words = input
    .toLowerCase()
    .split(/[,.]?\s+/)
    .map((word) => word.trim())
  const keywords = new Set<string>()

  // First pass: look for direct matches in our food relations and cuisine types
  for (const word of words) {
    if (foodRelations.has(word)) {
      keywords.add(word)
      // Add related terms
      foodRelations.get(word)?.forEach((term) => keywords.add(term))
    }
    if (cuisineTypes.has(word)) {
      keywords.add(word)
    }
  }

  // Second pass: look for compound terms (e.g., "stir fry")
  for (let i = 0; i < words.length - 1; i++) {
    const compound = words[i] + " " + words[i + 1]
    if (foodRelations.has(compound)) {
      keywords.add(compound)
      // Add related terms
      foodRelations.get(compound)?.forEach((term) => keywords.add(term))
    }
  }

  // If no keywords found, include original words that might be food-related
  if (keywords.size === 0) {
    words.forEach((word) => {
      // Add words that are likely food-related (not too short, not common words)
      if (word.length >= 3 && !isCommonWord(word)) {
        keywords.add(word)
      }
    })
  }

  return Array.from(keywords)
}

function isCommonWord(word: string): boolean {
  const commonWords = new Set([
    "the",
    "be",
    "to",
    "of",
    "and",
    "a",
    "in",
    "that",
    "have",
    "i",
    "it",
    "for",
    "not",
    "on",
    "with",
    "he",
    "as",
    "you",
    "do",
    "at",
    "this",
    "but",
    "his",
    "by",
    "from",
    "they",
    "we",
    "say",
    "her",
    "she",
    "or",
    "an",
    "will",
    "my",
    "one",
    "all",
    "would",
    "there",
    "their",
    "what",
    "so",
    "up",
    "out",
    "if",
    "about",
    "who",
    "get",
    "which",
    "go",
    "me",
    "when",
    "make",
    "can",
    "like",
    "time",
    "no",
    "just",
    "him",
    "know",
    "take",
    "people",
    "into",
    "year",
    "your",
    "good",
    "some",
    "could",
    "them",
    "see",
    "other",
    "than",
    "then",
    "now",
    "look",
    "only",
    "come",
    "its",
    "over",
    "think",
    "also",
    "back",
    "after",
    "use",
    "two",
    "how",
    "our",
    "work",
    "first",
    "well",
    "way",
    "even",
    "new",
    "want",
    "because",
    "any",
    "these",
    "give",
    "day",
    "most",
    "us",
    "need",
    "trying",
    "looking",
    "please",
    "help",
    "try",
    "much",
    "many",
    "lot",
    "very",
    "really",
    "something",
    "anything",
    "everything",
    "nothing",
    "find",
    "found",
    "want",
    "needs",
    "needed",
    "today",
    "tonight",
    "tomorrow",
    "yesterday",
    "week",
    "month",
    "year",
  ])
  return commonWords.has(word)
}

async function findRecipesByPreferences(
  preferences: string
): Promise<RecipeWithIngredients[]> {
  try {
    console.log("Starting recipe search with preferences:", preferences)

    // Extract keywords using OpenAI
    const searchTerms = await extractKeywordsWithAI(preferences)
    console.log("AI-extracted food keywords:", searchTerms)

    if (searchTerms.length === 0) {
      console.log("No valid search terms found")
      return []
    }

    // Build search conditions for each term
    const searchConditions = searchTerms.map((term) => {
      return sql`(
        LOWER(${recipesTable.title}) LIKE ${"%" + term + "%"} OR 
        LOWER(${recipesTable.description}) LIKE ${"%" + term + "%"} OR 
        LOWER(${ingredientsTable.name}) LIKE ${"%" + term + "%"}
      )`
    })

    // Combine conditions with OR
    const whereClause = sql.join(searchConditions, sql` OR `)

    // Search for recipes with ranking
    const recipes = await db
      .select({
        id: recipesTable.id,
        name: recipesTable.title,
        description: recipesTable.description,
        ingredients: sql<Array<{ name: string }>>`
          array_agg(DISTINCT jsonb_build_object('name', ${ingredientsTable.name}))
        `,
      })
      .from(recipesTable)
      .leftJoin(
        recipeIngredientsTable,
        eq(recipeIngredientsTable.recipeId, recipesTable.id)
      )
      .leftJoin(
        ingredientsTable,
        eq(ingredientsTable.id, recipeIngredientsTable.ingredientId)
      )
      .where(whereClause)
      .groupBy(recipesTable.id, recipesTable.title, recipesTable.description)
      .orderBy(
        sql`
        CASE 
          WHEN LOWER(${recipesTable.title}) LIKE ${searchTerms[0] + "%"} THEN 3
          WHEN LOWER(${recipesTable.title}) LIKE ${
          "%" + searchTerms[0] + "%"
        } THEN 2
          WHEN LOWER(${recipesTable.description}) LIKE ${
          "%" + searchTerms[0] + "%"
        } THEN 1
          ELSE 0
        END DESC
      `
      )
      .limit(10)

    console.log("Search results:", {
      termCount: searchTerms.length,
      searchTerms,
      recipeCount: recipes.length,
      topResults: recipes.map((r) => ({
        id: r.id,
        name: r.name,
      })),
    })

    return recipes
  } catch (error) {
    console.error("Error in findRecipesByPreferences:", {
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      preferences,
    })
    throw error
  }
}

export async function processChat(
  sessionId: string,
  messages: Message[]
): Promise<Message> {
  try {
    console.log("Starting processChat with sessionId:", sessionId)

    // Initialize or get chat state
    if (!chatStates.has(sessionId)) {
      console.log("Initializing new chat state for session:", sessionId)
      chatStates.set(sessionId, {
        messages: [],
        previouslySuggestedRecipes: [],
      })
    }

    const state = chatStates.get(sessionId)!
    state.messages = messages

    // Get the last user message
    const lastUserMessage = messages
      .slice()
      .reverse()
      .find((m) => m.role === "user")
    console.log("Last user message:", lastUserMessage?.content)

    if (lastUserMessage?.content) {
      console.log("Searching database for recipes...")
      const recipes = await findRecipesByPreferences(lastUserMessage.content)
      console.log(`Found ${recipes.length} total recipes`)

      // Filter out previously suggested recipes
      const availableRecipes = recipes.filter(
        (recipe) => !state.previouslySuggestedRecipes.includes(recipe.id)
      )
      console.log(
        `${availableRecipes.length} recipes available after filtering previously suggested`
      )

      if (availableRecipes.length > 0) {
        // Take the first available recipe
        const recipe = availableRecipes[0]
        state.previouslySuggestedRecipes.push(recipe.id)
        console.log("Selected recipe:", {
          id: recipe.id,
          name: recipe.name,
          ingredientCount: recipe.ingredients.length,
        })

        // Format the recipe suggestion with a link
        const recipeResponse = `Based on your interest in ${
          lastUserMessage.content
        }, I found a great recipe that you might enjoy:

[${recipe.name}](https://chopchoprecipes.com/recipes/${recipe.id})
${recipe.description || ""}

Would you like to see more recipes like this, or would you prefer something different?`

        return {
          role: "assistant",
          content: recipeResponse,
        }
      }
    }

    // If no recipes were found, suggest browsing the Explore page
    return {
      role: "assistant",
      content: `I couldn't find any recipes matching your preferences in our database. You can browse all our recipes on our [Explore Recipes](https://chopchoprecipes.com/recipes) page. Could you tell me more about what kinds of foods you enjoy?`,
    }
  } catch (error) {
    console.error("Error in processChat:", {
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      sessionId,
      lastMessage: messages[messages.length - 1]?.content,
    })
    return {
      role: "assistant",
      content:
        "I apologize, but I encountered an error while searching for recipes. You can browse all our recipes on our [Explore Recipes](https://chopchoprecipes.com/recipes) page. Would you like to try searching with different terms?",
    }
  }
}

File: RecipeApp.Api/Services/ChatService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using OpenAI_API;
using RecipeApp.Api.Data;
using RecipeApp.Api.Data.Entities;
using RecipeApp.Api.DTOs;
using System.Collections.Concurrent;

namespace RecipeApp.Api.Services;

public class ChatService : IChatService
{
    private readonly ApplicationDbContext _context;
    private readonly IConfiguration _configuration;
    private readonly ILogger<ChatService> _logger;
    private readonly OpenAIAPI _openAiApi;
    private readonly string _frontendUrl;

    // In-memory session state storage
    private static readonly ConcurrentDictionary<string, ChatState> ChatStates = new();

    public ChatService(ApplicationDbContext context, IConfiguration configuration, ILogger<ChatService> logger)
    {
        _context = context;
        _configuration = configuration;
        _logger = logger;
        var apiKey = configuration["OpenAI:ApiKey"] ?? throw new InvalidOperationException("OpenAI API key not configured");
        _openAiApi = new OpenAIAPI(apiKey);
        _frontendUrl = configuration["FrontendUrl"] ?? "https://chopchoprecipes.com";
    }

    public async Task<ChatMessageDto> ProcessChatAsync(string sessionId, List<ChatMessageDto> messages)
    {
        try
        {
            // Initialize or get chat state
            var state = ChatStates.GetOrAdd(sessionId, _ => new ChatState
            {
                Messages = new List<ChatMessageDto>(),
                PreviouslySuggestedRecipes = new List<int>(),
                LastSearchResults = new List<RecipeWithIngredients>(),
                LastSearchQuery = string.Empty
            });

            state.Messages = messages;

            // Get the last user message
            var lastUserMessage = messages
                .AsEnumerable()
                .Reverse()
                .FirstOrDefault(m => m.Role == "user");

            if (lastUserMessage == null || string.IsNullOrWhiteSpace(lastUserMessage.Content))
            {
                return new ChatMessageDto
                {
                    Role = "assistant",
                    Content = "I couldn't understand your message. Could you tell me what kinds of foods you enjoy?"
                };
            }

            // Check if user is asking for another recipe using OpenAI
            var requestType = await IsRequestingAnotherRecipeAsync(lastUserMessage.Content);

            if (requestType == "more" && state.LastSearchResults.Count > 0)
            {
                // Filter out previously suggested recipes
                var availableRecipes = state.LastSearchResults
                    .Where(r => !state.PreviouslySuggestedRecipes.Contains(r.Id))
                    .ToList();

                if (availableRecipes.Count > 0)
                {
                    var recipe = availableRecipes[0];
                    state.PreviouslySuggestedRecipes.Add(recipe.Id);

                    var recipeResponse = $@"Here's another recipe similar to what you're looking for:

[{recipe.Name}]({_frontendUrl}/recipes/{recipe.Id})
{recipe.Description ?? ""}

Would you like to see more recipes like this, something different, or would you like to [Explore all Recipes]({_frontendUrl}/recipes)?";

                    return new ChatMessageDto
                    {
                        Role = "assistant",
                        Content = recipeResponse
                    };
                }
                else
                {
                    return new ChatMessageDto
                    {
                        Role = "assistant",
                        Content = "I don't have any more recipes similar to your previous search. Would you like to try something different? Tell me what kind of recipe you're interested in."
                    };
                }
            }
            else if (requestType == "different")
            {
                return new ChatMessageDto
                {
                    Role = "assistant",
                    Content = "I'd be happy to help you find a different type of recipe. What kind of food are you in the mood for?"
                };
            }

            // Perform a new search
            var recipes = await FindRecipesByPreferencesAsync(lastUserMessage.Content);

            // Store the search results and query
            state.LastSearchResults = recipes;
            state.LastSearchQuery = lastUserMessage.Content;

            // Filter out previously suggested recipes
            var availableRecipesNew = recipes
                .Where(r => !state.PreviouslySuggestedRecipes.Contains(r.Id))
                .ToList();

            if (availableRecipesNew.Count > 0)
            {
                var recipe = availableRecipesNew[0];
                state.PreviouslySuggestedRecipes.Add(recipe.Id);

                var recipeResponse = $@"I found a great recipe that you might enjoy:

[{recipe.Name}]({_frontendUrl}/recipes/{recipe.Id})
{recipe.Description ?? ""}

Would you like to see more recipes like this, something different, or would you like to [Explore all Recipes]({_frontendUrl}/recipes)?";

                return new ChatMessageDto
                {
                    Role = "assistant",
                    Content = recipeResponse
                };
            }

            // If no recipes were found
            return new ChatMessageDto
            {
                Role = "assistant",
                Content = $"I couldn't find any recipes matching your preferences. Could you tell me more about what kinds of foods you enjoy?\nYou can also browse all our recipes on our [Explore Recipes]({_frontendUrl}/recipes) page."
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error in ProcessChatAsync");
            return new ChatMessageDto
            {
                Role = "assistant",
                Content = $"I apologize, but I encountered an error while searching for recipes. You can browse all our recipes on our [Explore Recipes]({_frontendUrl}/recipes) page. Would you like to try searching with different terms?"
            };
        }
    }

    private async Task<string> IsRequestingAnotherRecipeAsync(string message)
    {
        try
        {
            var chat = _openAiApi.Chat.CreateConversation();
            chat.AppendSystemMessage(@"You are a recipe assistant. Determine if the user wants:
1. More similar recipes (""more"")
2. Different types of recipes (""different"")
3. None of the above (""no"")

IMPORTANT: Initial food preferences or requests should be ""no"", not ""different"".

Respond with ONLY ""more"", ""different"", or ""no"".

Examples:
User: ""show me another one"" -> ""more""
User: ""I want to see more recipes like this"" -> ""more""
User: ""that looks good, what else do you have?"" -> ""more""
User: ""I want something different"" -> ""different""
User: ""show me something else instead"" -> ""different""
User: ""no thanks"" -> ""no""
User: ""I don't like that recipe"" -> ""different""
User: ""I like spicy food"" -> ""no""
User: ""I want sandwiches"" -> ""no""
User: ""show me pasta recipes"" -> ""no""
User: ""I'm looking for vegetarian dishes"" -> ""no""");
            chat.AppendUserInput(message);
            chat.RequestParameters.Temperature = 0.1;
            chat.RequestParameters.MaxTokens = 10;

            var response = await chat.GetResponseFromChatbotAsync();
            var answer = response?.ToLower().Trim() ?? "no";
            return answer == "more" || answer == "different" ? answer : "no";
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error in IsRequestingAnotherRecipeAsync");
            // Fall back to pattern matching
            var messageLower = message.ToLower();
            if (System.Text.RegularExpressions.Regex.IsMatch(messageLower, @"\b(show|give|get|want|see|try)\b.*\b(another|more)\b.*\b(recipe|one)\b"))
                return "more";
            if (System.Text.RegularExpressions.Regex.IsMatch(messageLower, @"\b(something|anything)\b.*\b(else|different)\b"))
                return "different";
            return "no";
        }
    }

    private async Task<List<RecipeWithIngredients>> FindRecipesByPreferencesAsync(string preferences)
    {
        try
        {
            // Extract keywords using OpenAI
            var searchTerms = await ExtractKeywordsWithAIAsync(preferences);

            if (searchTerms.Count == 0)
            {
                return new List<RecipeWithIngredients>();
            }

            // Build search query
            var query = _context.Recipes
                .Include(r => r.RecipeIngredients)
                    .ThenInclude(ri => ri.Ingredient)
                .AsQueryable();

            // Apply search filters - recipes must match at least one keyword
            var filteredQuery = query.Where(r =>
                searchTerms.Any(term =>
                    r.Title.ToLower().Contains(term) ||
                    (r.Description != null && r.Description.ToLower().Contains(term)) ||
                    r.RecipeIngredients.Any(ri => ri.Ingredient.Name.ToLower().Contains(term))
                )
            );

            // Get recipes first, then rank by keyword matches in memory
            var recipesData = await filteredQuery
                .Select(r => new RecipeWithIngredients
                {
                    Id = r.Id,
                    Name = r.Title,
                    Description = r.Description,
                    Ingredients = r.RecipeIngredients
                        .Select(ri => new IngredientInfo { Name = ri.Ingredient.Name })
                        .Distinct()
                        .ToList()
                })
                .ToListAsync();

            // Rank recipes by number of matching keywords (in memory)
            var recipesWithScores = recipesData.Select(r =>
            {
                var matchCount = searchTerms.Count(term =>
                    r.Name.ToLower().Contains(term) ||
                    (r.Description != null && r.Description.ToLower().Contains(term)) ||
                    r.Ingredients.Any(ing => ing.Name.ToLower().Contains(term))
                );
                return new { Recipe = r, MatchCount = matchCount };
            })
            .OrderByDescending(x => x.MatchCount)  // Recipes matching more keywords first
            .ThenBy(x => x.Recipe.Id)  // Then by ID for consistency
            .Take(10)
            .ToList();

            var recipes = recipesWithScores.Select(x => x.Recipe).ToList();

            // Log which keywords matched for debugging
            if (recipes.Count > 0)
            {
                _logger.LogInformation("[Recipe Search] Found {RecipeCount} recipes using {KeywordCount} keywords: [{Keywords}]", 
                    recipes.Count, searchTerms.Count, string.Join(", ", searchTerms));
                var topMatches = recipesWithScores.Take(3).Select(x => 
                    $"'{x.Recipe.Name}' ({x.MatchCount} matches)");
                _logger.LogInformation("[Recipe Search] Top matches: {TopMatches}", string.Join(", ", topMatches));
            }

            return recipes;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error in FindRecipesByPreferencesAsync");
            return new List<RecipeWithIngredients>();
        }
    }

    private async Task<List<string>> ExtractKeywordsWithAIAsync(string userInput)
    {
        try
        {
            // Security: Limit input length to prevent abuse
            if (string.IsNullOrWhiteSpace(userInput) || userInput.Length > 5000)
            {
                _logger.LogWarning("Invalid or too long user input for keyword extraction: {Length} characters", userInput?.Length ?? 0);
                return ExtractFoodKeywords(userInput ?? string.Empty);
            }

            var chat = _openAiApi.Chat.CreateConversation();
            chat.AppendSystemMessage(@"You are a culinary expert. Extract 5-10 food-related keywords from the user's input. 
CRITICAL: You MUST expand on the user's input - do NOT just echo back their exact words. Think of related ingredients, cuisines, dishes, cooking methods, flavors, and food categories.

Rules:
- Return AT LEAST 5 keywords, even if the user only mentions 1-2 words
- Include the user's input word(s) PLUS related culinary terms
- Focus on ingredients, cuisines, dishes, cooking methods, flavors, textures, and food categories
- Use ONLY singular forms (e.g., ""sandwich"" not ""sandwiches"")
- Return ONLY a comma-separated list of lowercase keywords, no explanations

Examples:
Input: ""pizza"" → Response: ""pizza, italian, cheese, tomato, dough, mozzarella, basil, marinara""
Input: ""chicken"" → Response: ""chicken, poultry, grilled, roasted, herbs, lemon, garlic, protein""
Input: ""spicy Asian food"" → Response: ""spicy, asian, stir-fry, noodle, curry, ginger, soy, chili, sesame""
Input: ""dessert"" → Response: ""dessert, sweet, sugar, chocolate, cake, fruit, vanilla, cream, pastry""");
            chat.AppendUserInput(userInput);
            chat.RequestParameters.Temperature = 0.5;  // Increased from 0.3 for more creativity
            chat.RequestParameters.MaxTokens = 100;  // Increased from 50 to allow more keywords

            var response = await chat.GetResponseFromChatbotAsync();
            
            // Log AI response for debugging
            _logger.LogInformation("[AI Keyword Extraction] User input: \"{UserInput}\"", userInput);
            _logger.LogInformation("[AI Keyword Extraction] AI response: \"{AiResponse}\"", response);
            
            if (!string.IsNullOrEmpty(response))
            {
                var keywords = response
                    .Split(',')
                    .Select(k => k.Trim().ToLower())
                    .Where(k => k.Length > 0)
                    .ToList();

                _logger.LogInformation("[AI Keyword Extraction] Extracted {KeywordCount} keywords: [{Keywords}]", 
                    keywords.Count, string.Join(", ", keywords));
                
                return keywords;
            }

            return ExtractFoodKeywords(userInput);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error in ExtractKeywordsWithAIAsync");
            return ExtractFoodKeywords(userInput);
        }
    }

    private List<string> ExtractFoodKeywords(string input)
    {
        // Basic fallback keyword extraction
        var words = input
            .ToLower()
            .Split(new[] { ' ', ',', '.' }, StringSplitOptions.RemoveEmptyEntries)
            .Select(w => w.Trim())
            .Where(w => w.Length >= 3)
            .ToList();

        return words;
    }

    private class ChatState
    {
        public List<ChatMessageDto> Messages { get; set; } = new();
        public List<int> PreviouslySuggestedRecipes { get; set; } = new();
        public List<RecipeWithIngredients> LastSearchResults { get; set; } = new();
        public string LastSearchQuery { get; set; } = string.Empty;
    }

    private class RecipeWithIngredients
    {
        public int Id { get; set; }
        public string Name { get; set; } = string.Empty;
        public string? Description { get; set; }
        public List<IngredientInfo> Ingredients { get; set; } = new();
    }

    private class IngredientInfo
    {
        public string Name { get; set; } = string.Empty;
    }
}
